### PR TITLE
Kj issue 11 lambda creation

### LINKED ISSUE #11: Write automation to create Lamdba resource for each developer
# User Story
As a developer, I want to write a script that creates individualized lambda functions, so I don't have to do it manually every time

# Time to complete
Estimated: 8 hours
Actual: 17 hours

# Acceptance Criteria
- [x] Create script that automatically generates a new lambda function
    - [x] Script should accept a name parameter for the lambda function

# Activities Documentation

| Date  | Activity | Time Spent |
| ------------- | ------------- | ------------- |
| 01/17/2022 | Researching how to create Lambdas using CLI | 1 hr  |
| 01//19/22 | Researching Cloudformation template + Lambdas  | 2 hr |
| 01/25/22 | Created basic Cloudformation template to create Lambda | 1 hour |
| 01/27/22 | Customized Cloudformation template and tested | 5 hours |
| 01/27/22 | Researched and created shell script and tested| 5 hours |
| 01/30/22 | Documentation including PR| 2 hours |
| 01/31/22 | Update PR | 1 hour |

# Pull Request
https://github.com/North-Seattle-College/ad440-winter2022-tuesday-repo/pull/36

### PR DESCRIPTION
**Pull request by KristinJJ**
Closes #11 

# Description 
As a developer, I want to write a script that creates individualized lambda functions, so I don't have to do it manually every time.

# Gallery
1. The script in the terminal
Screen Shot 2022-01-31 at 1 43 02 AM

2. CloudFormation Stack with automatically generated name
Screen Shot 2022-01-31 at 1 45 34 AM

3. Lambda function with automatically generated name
Screen Shot 2022-01-31 at 1 46 21 AM

4. Configure a test event
Screen Shot 2022-01-31 at 1 00 08 PM

5. Choose the "hello-world" template, name the test, and save it
Screen Shot 2022-01-31 at 12 56 29 PM

6. How to find your created test event
Screen Shot 2022-01-31 at 12 56 46 PM

# Tested
macOS in bash shell and VS Code

# Testing
### Creating the cloudformation stack and lambda function
- [ ] Navigate to the makeLambdaFunction folder
- [ ] Run `./makeLambdaFunction.sh`
- [ ] Follow the prompts to enter your first and last initial (see image 1 in the description above)
- [ ] View the automatically generated names for Lambda function and CloudFormation Stack
- [ ] Enter `y` to confirm creation of Lambda function and CloudFormation Stack, or enter `n` to exit the program
- [ ] If `y` is entered, CloudFormation template will be deployed
- [ ] After stack has been created, view `Successfully created/updated stack`
- [ ] Navigate to AWS console
- [ ] Search CloudFormation to view created stacks, and confirm newly created stack is present (see image 2 in the description above)
- [ ] Search Lambda to view Lambda functions, and confirm newly created Lambda function is present (see image 3 in the description above)
### Testing the lambda function
- [ ] To verify that the Lambda function works, click on the newly created Lambda function
- [ ] To create a test event, click on the orange Test button and choose "Configure test event" (see image 4 in the description above)
- [ ] Choose "Create new test event" (see image 5 in the description above)
- [ ] From the Event template dropdown, choose the "hello-world" template
- [ ] Enter your Event name
- [ ] Click the Create button
- [ ] To run your newly created test event, click on the orange Test button and choose your test event (see image 6 in the description above)
- [ ] If everything is working, in the Execution results tab you should see
`{
  "statusCode": 200,
  "body": "\"Hello from Lambda!\""
}`

### CODE REVIEW

File: devops/makeLambdaFunction/makeLambdaFunction.yml
Comment (dansjack) [inline devops/makeLambdaFunction/makeLambdaFunction.yml:15]: I was curious about how we might be able to pass a file here and found this SO post that might be useful in the future: 

[https://stackoverflow.com/questions/54876833/upload-local-file-on-parameter-zipfile-awslambdafunction](https://stackoverflow.com/questions/54876833/upload-local-file-on-parameter-zipfile-awslambdafunction)

File: devops/makeLambdaFunction/makeLambdaFunction.yml
Comment (dansjack) [inline devops/makeLambdaFunction/makeLambdaFunction.yml:1]: I see that you've written your automation in `yml` while I wrote mine in `json`. We should decide on a consistent format moving forward for all of these CloudFormation templates. I'm happy to stick with `yml` unless other devops folks have a preference!

**Review by dansjack — CHANGES_REQUESTED** [review]
Testing steps worked well! The script created the resources and I tested that the lambda function worked.

image

I'm only requesting changes because of the `index.html` file included in the PR. If that's necessary, please let me know. Otherwise, when it's removed, I'll approve!

File: devops/index.html
Comment (KristinJJ) [reply devops/index.html]: Nope, it does not. Removing this file from the PR. Thanks @dansjack !

File: devops/makeLambdaFunction/makeLambdaFunction.yml
Comment (KristinJJ) [reply devops/makeLambdaFunction/makeLambdaFunction.yml:15]: Nice, thanks for the resource @dansjack !

File: devops/makeLambdaFunction/makeLambdaFunction.yml
Comment (KristinJJ) [reply devops/makeLambdaFunction/makeLambdaFunction.yml:1]: Good point @dansjack! I personally feel like yml is easier to read because of less syntax, so it can be easier to see what is going on while configuring the template, but I am also happy to convert to json format in the future.

Comment (dansjack) [reply devops/makeLambdaFunction/makeLambdaFunction.yml:1]: Sounds good! Lets go with yml

**Review by dansjack — APPROVED** [review]
Great work!

**Review by aaronallenCS — APPROVED** [review]
Looks good to me!

**Review by Annelisebx — COMMENTED** [review]
image
created lambda function!

**Comment by Jflands23** [comment]
Went over PR after it was merged but still tested everything out. works great!

**Pull request merged**
